cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_node_helpers.cpp

```diff
@@ -0,0 +1,131 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/decision_forest/backend/gpu/train_node_helpers.hpp"
+#include "oneapi/dal/detail/profiler.hpp"
+#include "oneapi/dal/algo/decision_forest/backend/gpu/train_misc_structs.hpp"
+#include "oneapi/dal/algo/decision_forest/backend/gpu/train_helpers.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+namespace oneapi::dal::decision_forest::backend {
+
+using alloc = sycl::usm::alloc;
+
+namespace de = dal::detail;
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+template <typename Index>
+sycl::event node_group_list<Index>::filter_internal(const node_list_t& node_list,
+                                                    const bk::event_vector& deps) {
+    ONEDAL_PROFILER_TASK(split_node_list_on_groups_by_size, queue_);
+
+    ONEDAL_ASSERT(node_list.state_is_valid());
+    ONEDAL_ASSERT(this.state_is_valid());
+
+    const Index node_prop_count = node_t::get_prop_count(); // num of split attributes for node
+
+    const Index* node_list_ptr = node_list.get_list().get_data();
+    const Index node_count = node_list.get_count();
+
+    auto node_groups = this->get_list();
+    constexpr Index group_count = node_group_list_t::get_count();
+    constexpr Index group_prop_count = node_group_t::get_prop_count();
+
+    const Index* node_groups_bound_ptr = this->get_bound_list().get_data();
+    Index* node_groups_ptr = node_groups.get_mutable_data();
+    Index* node_indices_ptr = this->get_node_indices_list().get_mutable_data();
+
+    Index max_sbg_size = bk::device_max_sg_size(queue_);
+    Index local_size = std::max(Index(max_sbg_size * group_count), min_local_size_);
+    ONEDAL_ASSERT(bk::device_max_wg_size(queue_) >= local_size);
+
+    const sycl::nd_range<1> nd_range = bk::make_multiple_nd_range_1d(local_size, local_size);
+
+    auto event = queue_.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        local_accessor_rw_t<Index> local_buf(local_size, cgh);
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
+            auto sbg = item.get_sub_group();
+            const Index sub_group_id = sbg.get_group_id();
+
+            const Index local_id = sbg.get_local_id();
+            const Index local_size = sbg.get_local_range()[0];
+
+            Index* local_buf_ptr = local_buf.get_pointer().get();
+
+            Index bucket_count = 0;
```

Comment:
Not sure it should be Index(0) but it would be good to have the same style everywhere

Reply:
i removed Index(<int_const>) in other places